xmppserver/src/test/java/org/jivesoftware/openfire/session/LocalOutgoingServerSessionTest.java

```diff
@@ -0,0 +1,335 @@
+        final ConnectionManagerImpl connectionManager = Fixtures.mockConnectionManager();
+        final ConnectionListener connectionListener = Fixtures.mockConnectionListener();
+        doReturn(connectionConfiguration).when(connectionListener).generateConnectionConfiguration();
+        doReturn(Set.of(connectionListener)).when(connectionManager).getListeners(any(ConnectionType.class));
+        doReturn(connectionListener).when(connectionManager).getListener(any(ConnectionType.class), anyBoolean());
+        doReturn(connectionManager).when(xmppServer).getConnectionManager();
+    }
+
+    @Before
+    public void setUp() throws Exception
+    {
+        remoteServerDummy = new RemoteServerDummy();
+        remoteServerDummy.open();
+
+        Fixtures.clearExistingProperties();
+        DNSUtil.setDnsOverride(Map.of(RemoteServerDummy.XMPP_DOMAIN, new DNSUtil.HostAddress("localhost", remoteServerDummy.getPort(), false)));
+    }
+
+    @After
+    public void tearDown() throws Exception
+    {
+        DNSUtil.setDnsOverride(null);
+
+        if (remoteServerDummy != null) {
+            remoteServerDummy.close();
+            remoteServerDummy = null;
+        }
+
+        Fixtures.clearExistingProperties();
+    }
+
+    /**
+     * Verifies that an authenticated and encrypted connection can be established when:
+     *
+     * <ul>
+     * <li>the remote server identifies itself using a certificate chain that is valid and uses a trusted CA.
+     * <li>Dialback authentication is disabled
+     * </ul>
+     *
+     * Verifying that the connection is encrypted, and authenticated NOT using DIALBACK (implying that SASL EXTERNAL is used).
+     */
+    @Test
+    public void testOutboundS2S_SaslExternal_PeerUsesSignedCert() throws Exception
+    {
+        final TrustStore trustStore = XMPPServer.getInstance().getCertificateStoreManager().getTrustStore(ConnectionType.SOCKET_S2S);
+        try {
+            // Setup test fixture.
+            JiveGlobals.setProperty(ConnectionSettings.Server.DIALBACK_ENABLED, "false");
+            remoteServerDummy.preparePKIX();
+            final X509Certificate[] chain = remoteServerDummy.getGeneratedPKIX().getCertificateChain();
+            final X509Certificate caCert = chain[chain.length-1];
+            trustStore.installCertificate("unit-test", KeystoreTestUtils.toPemFormat(caCert));
+
+            final DomainPair domainPair = new DomainPair(Fixtures.XMPP_DOMAIN, RemoteServerDummy.XMPP_DOMAIN);
+            final int port = remoteServerDummy.getPort();
+
+            // Execute system under test.
+            final LocalOutgoingServerSession result = LocalOutgoingServerSession.createOutgoingSession(domainPair, port);
+
+            // Verify results.
+            assertNotNull(result);
+            assertFalse(result.isClosed());
+            assertTrue(result.isEncrypted());
+            assertFalse(result.isUsingServerDialback());
```

Comment:
More a style thing, but more recently I've become a big fan of AssertJ - https://assertj.github.io/doc/

The above four assertions could then be written more fluently as
```java
assertThat(result).isNotNull();
assertThat(result.isClosed()).isFalse();
assertThat(result.isEncrypted).isTrue();
assertThat(result.isUsingDialback).isFalse();
```
Though the real benefit comes when using things like reflective comparisons on lists, etc.

Reply:
I myself have no strong preference for one style or the other. Currently, I'm struggling to get the changes in this PR to be correct. I fear that moving away from my style-based comfort zone (even for an improvement) might overload my poor brain. :)